### PR TITLE
lib: Add lib.maintainer-groups

### PR DESCRIPTION
**Pull request by jtojnar**
There are many people interested in being responsible for core desktop packages but adding themselves to each one of them is a chore. This PR adds an indirection simplifying this use case.

### CODE REVIEW

**Comment by jtojnar** [comment]
I do not like having to do `.maintainers` but I still want to have attrsets so we do not have to repeat https://github.com/NixOS/nixpkgs/pull/36119

**Comment by worldofpeace** [comment]
Makes sense, I wouldn't have to have a `pkgs.pantheon.maintainers` or `pkgs.gnome3.maintainers` anymore (that doesn't make sense at the `pkgs` level anyway). 
I was thinking of adding myself to glib and gtk3 today but this idea returned to me and thought "I'd rather take the time to have this". Guess our idea was floating around again :smile:

**Comment by jtojnar** [comment]
Yeah, `colord-gtk` is package without `meta.maintainers` but I did not want to add myself directly.

**Comment by worldofpeace** [comment]
So just an idea dump, I think I might've talked about having something where this could map to [NixOS Teams](https://github.com/orgs/NixOS/teams) so when there's official "working groups" ofborg could request their review. Autolabeing issues for the groups would be nice also, and it would help with people figuring out who's the best person/s to reach on most things. But that's probably an idea that would needs the organizational effort of an RFC.

**Comment by jonringer** [comment]
I like the idea of having the maintainer-groups being centralized to a single location. Hard to know which packages have maintainers like mesa, gtk, etc

**Comment by aanderse** [comment]
This is much healthier for nixpkgs core/important packages and will make keeping things up to date so much easier. I love it!

**Comment by jtojnar** [comment]
Maybe renaming it to `lib.teams.freedesktop.members` would be little bit nicer to type. 

Eventually, we would like to be able to just add a group directly to the `meta.maintainers` list but that would require changes to OfBorg and Hydra. (Though it should not be a difficult change, just merging and deduplicatiom.) We can leave that for future PRs when this proves useful.

One disadvantage of the current implementation is that it is not possible to find out all packages maintained by a group. That would be useful for `maintainers/scripts/update.nix`. Maybe we should go directly to phase two as described above.

File: pkgs/tools/misc/colord/default.nix [hunk 1]
```diff
@@ -101,7 +101,7 @@ stdenv.mkDerivation rec {
     description = "System service to manage, install and generate color profiles to accurately color manage input and output devices";
     homepage = https://www.freedesktop.org/software/colord/;
     license = licenses.lgpl2Plus;
-    maintainers = [ maintainers.marcweber ];
```
Comment (grahamc) [inline pkgs/tools/misc/colord/default.nix]: maybe should be called `teams` instead?

**Comment by worldofpeace** [comment]
@jtojnar So the idea here is that teams are special groups of maintainers that have shared responsibility, ownership, and accountability. They've like maintainers but they operate more collectively, and there's a usually an ongoing internal discourse and larger infrastructures to be accountable for.

I don't think this makes `teams` a different meta attribute, just a different type of maintainer. 
And maintainers list should  should be taught groups as you said.

I do think this could be useful right away, it's just when ofborg requests reviews it'll be as individuals.
But people can still read the source description of `teams`. Phase two as described is a good idea.

**Comment by worldofpeace** [comment]
Though it is a fundamental thing, should we land more gracefully? It's more difficult to organize changes like this that have many pieces. So if something is committed first it'll be easier down the line.

**Comment by jtojnar** [comment]
Here is an example how the first-class teams could be handled by ofborg: https://github.com/NixOS/ofborg/pull/421

**Comment by jtojnar** [comment]
> Though it is a fundamental thing, should we land more gracefully? It's more difficult to organize changes like this that have many pieces. So if something is committed first it'll be easier down the line.

On the other hand, not having first-class teams at first would mean having to modify every instance of legacy use. And it could possibly hinder adoption because of people waiting for a better implementation.

File: maintainers/team-list.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    name = {
+      # Required
+      members = [ maintainer1 maintainer2 ];
+      scope = "Maintain foo packages.";
+    };
+
+  where
+
+  - `members` is the list of maintainers belonging to the group,
+  - `scope` describes the scope of the group.
```
Comment (grahamc) [inline maintainers/team-list.nix]:
```suggestion
  - `scope` describes the scope of the group,
  - `type` "interest-group" or  "business",
  - `membership` "open" or "closed"; to join a closed group, existing members must approve.
```

**Review by grahamc — COMMENTED** [review]
I'd like to merge this. I know of a company which would like to have a maintainer group for themselves to attach to packages they as an entity maintain. I added a couple of new fileds, type and membership, because a business-maintained group should probably not casually get more users added to it.

What do y'all think?

@ofborg eval

File: maintainers/team-list.nix (hunk 2 above)
Comment (jtojnar) [reply maintainers/team-list.nix]: Sounds reasonable.

**Comment by grahamc** [comment]
@ofborg eval

File: maintainers/team-list.nix (hunk 2 above)
Comment (jtojnar) [reply maintainers/team-list.nix]: Relevant IRC discussion https://logs.nix.samueldr.com/nixos-dev/2020-03-09#1583795904-1583796772;

**Comment by jonringer** [comment]
I'm just a little concerned about the following scenario:
```
meta = with lib; {
  ...
  maintainers = with maintainers; with teams; [ jonringer grahamc ] ++ freedesktop.members;
  ...
```

hard to know where the attr's are coming from, I guess it could be changed to:
```
meta = with lib; {
  ...
  maintainers = with maintainers; [ jonringer grahamc ] ++ teams.freedesktop.members;
  ...
```
similar to the PR, and that's fairly acceptable.

Knowing what attr's are being introduce by a `with` clause has been a consistent user-pain for nix, even outside of this PR. So I won't hold it against it.

**Review by jonringer — APPROVED** [review]

**Pull request merged**